src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,395 @@
+            {
+                return;
+            }
+
+            var formatStringLiteralExpressionSyntax = TryGetFormatStringLiteralExpressionSyntax(
+                context.SemanticModel,
+                arguments,
+                parameters);
+
+            if (formatStringLiteralExpressionSyntax == null)
+            { 
+                return;
+            }
+
+            if (parameters.Length == 0)
+            {
+                return;
+            }
+
+            var hasIFormatProvider = parameters[0].Type.Equals(formatProviderType);
+
+            // We know the format string parameter exists so numberOfArguments is at least one,
+            // and at least 2 if there is also an IFormatProvider.  The result can be zero if
+            // calling string.Format("String only with no placeholders"), where there is an
+            // empty params array.
+            var numberOfPlaceholderArguments = numberOfArguments - (hasIFormatProvider ? 2 : 1);
+
+            var formatString = formatStringLiteralExpressionSyntax.ToString();
+
+            if (FormatCallWorksAtRuntime(formatString, numberOfPlaceholderArguments))
+            {
+                return;
+            }
+
+            ValidateAndReportDiagnostic(context, numberOfPlaceholderArguments,
+                formatString, formatStringLiteralExpressionSyntax.SpanStart);
+        }
+
+        private bool IsValidFormatMethod(ISyntaxFactsService syntaxFacts, SyntaxNode expression)
+        {
+            // When calling string.Format(...), the expression will be MemberAccessExpressionSyntax
+            if (syntaxFacts.IsSimpleMemberAccessExpression(expression))
+            {
+                var nameOfMemberAccessExpression = syntaxFacts.GetNameOfMemberAccessExpression(expression);
+                return !syntaxFacts.IsGenericName(nameOfMemberAccessExpression)
+                    && syntaxFacts.GetIdentifierOfSimpleName(nameOfMemberAccessExpression).
+                    ValueText.Equals(nameof(string.Format));
+            }
+
+            // When using static System.String and calling Format(...), the expression will be
+            // IdentifierNameSyntax
+            if (syntaxFacts.IsIdentifierName(expression))
+            {
+                return syntaxFacts.GetIdentifierOfSimpleName(expression).ValueText.Equals(
+                    nameof(string.Format));
+            }
+
+            return false;
+        }
+
+        private bool ArgsIsArrayOfReferenceTypes(
+            SemanticModel semanticModel,
+            SeparatedSyntaxList<SyntaxNode> arguments,
+            ImmutableArray<IParameterSymbol> parameters)
+        {
+            var argsArgumentType = TryGetArgsArgumentType(semanticModel, arguments, parameters);
+            if (argsArgumentType == null)
+            {
+                return false;
+            }
+
+            if (argsArgumentType.TypeKind != TypeKind.Array)
```

Comment:
consider simplifying entirely to "return argsArgumentType is IArrayTypeSymbol arrayType && arrayType.ElementType.IsReferenceType;"